File: apps/go/internal/chapter07/cinemadata_test.go
package chapter07_test

import (
	"bytes"
	"math"
	"os"
	"path/filepath"
	"testing"

	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter02"
	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter07"
	"github.com/k2works/getting-started-machine-learning/apps/go/internal/dataset"
)

// requireData は学習データが無ければテストを飛ばす。
func requireData(t *testing.T) string {
	t.Helper()

	csvFile := filepath.Join(dataset.Current(), "cinema.csv")
	if _, err := os.Stat(csvFile); err != nil {
		t.Skip("学習データ cinema.csv が配置されていない（gulp data:setup）")
	}

	return csvFile
}

func TestCinemaData(t *testing.T) {
	t.Parallel()

	t.Run("外れ値を 1 件除いて 99 件になる", func(t *testing.T) {
		t.Parallel()

		table, err := chapter02.LoadTable(requireData(t))
		if err != nil {
			t.Fatalf("LoadTable() でエラー: %v", err)
		}

		cleaned, err := chapter07.RemoveOutliers(table)
		if err != nil {
			t.Fatalf("RemoveOutliers() でエラー: %v", err)
		}

		if len(table.Rows) != 100 || len(cleaned.Rows) != 99 {
			t.Errorf("件数 = %d と %d, want 100 と 99", len(table.Rows), len(cleaned.Rows))
		}
	})

	t.Run("テストデータの決定係数が 0.6591 になる", func(t *testing.T) {
		t.Parallel()

		split, err := chapter07.PrepareCinema(requireData(t), 0.2, 0)
		if err != nil {
			t.Fatalf("PrepareCinema() でエラー: %v", err)
		}

		model, err := chapter07.Fit(split.XTrain, split.TTrain)
		if err != nil {
			t.Fatalf("Fit() でエラー: %v", err)
		}

		predictions, err := model.PredictAll(split.XTest)
		if err != nil {
			t.Fatalf("PredictAll() でエラー: %v", err)
		}

		got, err := chapter07.R2Score(split.TTest, predictions)
		if err != nil {
			t.Fatalf("R2Score() でエラー: %v", err)
		}

		if math.Abs(got-0.6591) > 5e-5 {
			t.Errorf("決定係数 = %v, want 0.6591", got)
		}
	})

	t.Run("実行すると係数と評価指標を表示する", func(t *testing.T) {
		t.Parallel()

		_ = requireData(t)

		var out bytes.Buffer
		if err := chapter07.Run(&out); err != nil {
			t.Fatalf("Run() でエラー: %v", err)
		}

		// Go の math/rand は分け方がほかの言語版と違うので、係数も評価指標も一致しない
		want := "データ件数: 100\n" +
			"外れ値を除いた件数: 99\n" +
			"訓練データ: 79 件, テストデータ: 20 件\n" +
			"切片: 6392.98\n" +
			"係数: SNS1=1.2610, SNS2=0.5934, actor=0.2577, original=201.1699\n" +
			"テストデータの評価: R2=0.6591, MAE=342.40, RMSE=409.68\n" +
			"単回帰（SNS2 のみ・gonum）: 切片=8979.41, 係数=1.3074\n"
		if out.String() != want {
			t.Errorf("出力 = %q, want %q", out.String(), want)
		}
	})
}

File: apps/clojure/test/getting_started_ml/cinema_data_test.clj
(ns getting-started-ml.cinema-data-test
  "実データ（cinema.csv）で線形回帰を学習するテスト。学習データが無ければスキップする。"
  (:require [clojure.string :as str]
            [clojure.test :refer [deftest is]]
            [getting-started-ml.chapter02 :as chapter02]
            [getting-started-ml.chapter07 :as ch]
            [getting-started-ml.chapter07.metrics :as metrics]
            [getting-started-ml.chapter07.tribuo :as tribuo]
            [getting-started-ml.dataset :as dataset])
  (:import [org.tribuo.regression.slm LARSTrainer SLMTrainer]))

(def ^:private csv-file (delay (str (dataset/dir) "/cinema.csv")))

(defn- data?
  "学習データがあるかどうかを返す。無ければ理由を標準エラーに出す。"
  []
  (or (.exists (java.io.File. ^String @csv-file))
      (binding [*out* *err*]
        (println "学習データ cinema.csv が配置されていない（gulp data:setup）のでスキップする")
        false)))

(deftest 実データから外れ値を一件取り除く
  (when (data?)
    (let [table (chapter02/load-table @csv-file)]
      (is (= 100 (count (:rows table))))
      (is (= 99 (count (:rows (ch/remove-outliers table))))))))

(deftest 実データで自作のモデルとTribuoのSLMTrainerとLARSTrainerのR2が一致する
  (when (data?)
    (let [split (ch/prepare-cinema @csv-file 0.2 0)
          columns ch/feature-columns
          mine (metrics/r2-score (:t-test split)
                                 (ch/predict (ch/fit (:x-train split) (:t-train split) columns)
                                             (:x-test split)))]
      (doseq [trainer [(SLMTrainer. true) (LARSTrainer.)]]
        (let [model (tribuo/train trainer (:x-train split) (:t-train split) columns)
              theirs (metrics/r2-score (:t-test split)
                                       (tribuo/predict model (:x-test split) columns))]
          (is (< (abs (- theirs mine)) 1e-9) (str (.getSimpleName (class trainer)))))))))

(deftest 実行すると学習した係数と評価指標を表示する
  (when (data?)
    (is (= ["データ件数: 100"
            "外れ値を除いた件数: 99"
            "訓練データ: 79 件, テストデータ: 20 件"
            "切片: 6114.60"
            "係数: SNS1=1.3804, SNS2=0.5218, actor=0.2900, original=208.8827"
            "テストデータの評価: R2=0.6184, MAE=302.20, RMSE=376.14"]
           (str/split-lines (with-out-str (ch/run)))))))
